akka-actor/src/main/scala/akka/actor/CoordinatedShutdown.scala

```diff
@@ -354,16 +354,75 @@ final class CoordinatedShutdown private[akka] (
     system: ExtendedActorSystem,
     phases: Map[String, CoordinatedShutdown.Phase])
     extends Extension {
-  import CoordinatedShutdown.Reason
-  import CoordinatedShutdown.UnknownReason
+  import CoordinatedShutdown.{ Reason, UnknownReason }
 
   /** INTERNAL API */
   private[akka] val log = Logging(system, getClass)
   private val knownPhases = phases.keySet ++ phases.values.flatMap(_.dependsOn)
 
   /** INTERNAL API */
   private[akka] val orderedPhases = CoordinatedShutdown.topologicalSort(phases)
-  private val tasks = new ConcurrentHashMap[String, Vector[(String, () => Future[Done])]]
+
+  private object tasks {
+    private val registeredTasks = new ConcurrentHashMap[String, Map[() => Future[Done], Vector[String]]]()
+
+    def get(phase: String): Option[Map[() => Future[Done], Vector[String]]] = Option(registeredTasks.get(phase))
+
+    def totalDuration(): FiniteDuration = {
+      import akka.util.ccompat.JavaConverters._
+      registeredTasks.keySet.asScala.foldLeft(Duration.Zero) {
+        case (acc, phase) =>
+          acc + timeout(phase)
+      }
+    }
+
+    def register(phase: String)(task: () => Future[Done], name: String): Cancellable = {
+      registeredTasks.merge(
+        phase,
+        Map(task -> Vector(name)),
+        (t, u) =>
+          t ++ u.map {
+            case (k, names) =>
+              k -> (t.get(k).toVector.flatten ++ names)
+          })
+      new Cancellable {
+        private val p = Promise[Unit]()
+
+        override def cancel(): Boolean = {
+          if (p.isCompleted) { // This ensures that multiple calls to `cancel` don't unregister another task with the same name
+            false
+          } else {
+            unregister(phase)(task, name) && p.trySuccess {
```

Comment:
isn't that a concurrency race condition between the `p` check` and `unregister`?
maybe `unregister` is idempotent and works fine for multiple calls and the purpose of `p` is only to avoid double logging?

what is the name used for now? Previously it was only used for logging (as far as I recall). Is it used as some kind of identifier now?

Reply:
Each task maintains a vector of the names under which it has been registered in the concurrent hashmap. When `unregister(phase)(task, name)` is called, we look in `task`'s vector for a matching name, and if we find one, we delete one copy of that name (and then if the resulting vector is empty, discard `task` from the mapping).

The purpose of `p` is, if the same task has been registered twice under the same name, to prevent the owner of just one Cancellable from cancelling both itself and the other.

There is a short race condition here. It could be resolved by moving `p.trySuccess(...)` before the call to `unregister`; one reason I put it after was the (remote) chance that the update operation on the concurrent hashmap failed with an exception; this would leave the Cancellable unable to be cancelled. I'm not seeing a solution that guarantees all of
* idempotence of `.cancel()
* *if* `.cancel()` returns `true` *then* this copy of task was successfully removed;
* if cancellation fails with an exception then cancellation can be retried fruitfully